pages/en/lb4/Mixin.md

```diff
@@ -0,0 +1,119 @@
+---
+lang: en
+title: 'Mixin'
+keywords: LoopBack 4.0, LoopBack 4
+tags:
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Mixin.html
+summary:
+---
+
+It is a commonly used JavaScript/TypeScript strategy to extend a class with new properties and methods.
+
+*question: does our mixinBuilder allow interface extends classes?*
+
+## Mixin Builder
+
+LoopBack provides a mixin builder function `MixinBuilder` to easily extend a class 
+with one or multiple mixins,
+you can call it to define the new extended class with fluent syntaxed API like:
+
+```ts
+import {MixiBuilder} from '@loopback/repository';
+import {BaseClass} from '<path_exports_BaseClass>';
+import {FooMixin, BarMixin} from '<path_exports_Mixins>';
+
+let newClass = MixinBuilder.mix(baseClass).with(FooMixin, BarMixin);
```

Comment:
This example has an important problem: it's assigning a class constructor to a variable `newClass`. The usual convention is to start class names with a capital letter, e.g. `NewClass`. However, because we are using `let` + assignment instead of `class`, the name `NewClass` will trigger a tslint error about violating naming conventions for variables.

IMO, we should not promote MixinBuilder and tell people to compose mixins directly via regular JS/TS function invocation and class inheritance:

```ts
class NewClass extends FooMixin(BarMixin(baseClass)) {
}
```

Another important difference to consider:

In my proposed approach, the javascript runtime see a correct class name (`NewClass`) and `NewClass.toString()` operator returns `class NewClass...`:

```
> NewClass.name
"NewClass"

> NewClass.toString()
"class NewClass extends Mixin1(Mixin2(BaseClass)) {
        }"
```

Compare it with the result of `MixinBuilder.mix().with()`, which incorrectly sets the constructor name to `BaseClass` and where `newClass.toString()` returns a fully mixed class definition that makes it difficult to see what have been added by what mixin:

```
> newClass.name
"BaseClass"

> newClass.toString()
"class extends superClass {
        constructor() {
            super(...arguments);
            this.mixinProp2 = 'mixinProp2';
        }
        static staticMixinMethod2() {
            return 'mixin2.static';
        }
        mixinMethod2() {
            return 'mixin2';
        }
    }"
```

I am using the following unit-test code to verify my assumptions and produce sample output: https://github.com/strongloop/loopback-next/blob/568c6e2e7242949e1e5ece040bc3a54d83ff1411/packages/repository/test/unit/mixin/mixin.ts#L53

cc @raymondfeng as the author of MixinBuilder

Reply:
>we should not promote MixinBuilder and tell people to compose mixins directly via regular JS/TS function invocation and class inheritance:

I wonder if one of the objectives of our MixinBuilder is to compose mixins. If that's the case, IMO shouldn't the constructor name and class def be preserved by MixinBuilder? (unless there is a reason I'm not aware of)

Reply:
@raymondfeng Could you take a look of comment https://github.com/strongloop/loopback.io/pull/491/files#r145687471? Thanks!

My understanding of `MixinBuilder` is it provides a prettier syntax to compose mixins to a class, but @bajtos has a good point in comment above.

Reply:
I borrowed the idea from http://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/

Reply:
FYI 
> Compare it with the result of MixinBuilder.mix().with(), which incorrectly sets the constructor name to BaseClass and where newClass.toString() returns a fully mixed class definition that makes it difficult to see what have been added by what mixin:

This could be resolved by defining `newClass` as 
`newClass = class extends MixinBuilder.mix(BaseClass).with(Mixin1, Mixin2) {};`

Then print out:
```ts
console.log(newClass.name);
// newClass
console.log(newClass.toString());
// class extends _1.MixinBuilder.mix(BaseClass).with(Mixin1, Mixin2) {}
```

The problem of `MixinBuilder` from my test is: function `with()` doesn't set its returned type as a class extends the `BaseClass`:
[click to see test example](https://github.com/strongloop/loopback-next/blob/master/packages/repository/test/unit/repository-mixin/repository-mixin.test.ts#L60)
```ts
let AppWithRepoMixin = class extends MixinBuilder.mix(Application).with(RepositoryMixin){};

const myApp = new AppWithRepoMixin();
const boundComponentsBefore = myApp.find('components.*').map(b => b.key);
``` 
fails when compiling, because compiler doesn't know `myApp` is an instance of `Application`, therefore doesn't know it has a prototype function `find()` returning an `Array`. The code calls `map` with `any` not `Array` and fails.

This looks a stop for people wanting to use it...

I am going to defer writing doc for `MixinBuilder` and create an issue for it. I still love the fluent api and would like to dig more :star:

Reply:
> I am going to defer writing doc for MixinBuilder and create an issue for it. I still love the fluent api and would like to dig more ⭐️

@jannyHou could you please paste a link to that new issue here for posterity?

Reply:
@bajtos sure, forgot to link it here: https://github.com/strongloop/loopback-next/issues/673